File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaInstanceNode.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjects;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaReferenceTypes;

public abstract class UaInstanceNode extends UaNode {
    private UaModellingRule modellingRule;

    public UaInstanceNode(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName)
    {
        super(nodeId, browseName, displayName);
        modellingRule = UaModellingRule.None;
    }

    public UaModellingRule modellingRule() {
        return modellingRule;
    }

    public void setModellingRule(UaModellingRule modellingRule) throws UaRuntimeException {
        if (UaModellingRule.None != this.modellingRule) throw new UaRuntimeException(StatusCodes.Bad_AlreadyExists);
        if (UaModellingRule.None == modellingRule) return;

        this.modellingRule = modellingRule;

        if (UaModellingRule.Mandatory == modellingRule)
        {
            addReference(new UaReference(UaObjects.ModellingRule_Mandatory, UaReferenceTypes.HasModellingRule, true));
        }
    }
}

File: opcua-webservice-node/src/server/addressspace/nodes/UaInstanceNode.ts
import { StatusCodes } from "opcua-webapi";
import { UaError, UaLocalizedText, UaNodeId, makeUaStatusCode } from "opcua-webapi-ts";
import { UaNode,UaReference,UaModellingRule, UaObjects, UaReferenceTypes } from ".";

export abstract class UaInstanceNode extends UaNode {
    private _modellingRule: UaModellingRule;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText
    ) {
        super(nodeId, browseName, displayName);
        this._modellingRule = UaModellingRule.None;
    }

    public get modellingRule(): UaModellingRule {
        return this._modellingRule;
    }

    public setModellingRule(modellingRule: UaModellingRule): void {
        if (this._modellingRule !== UaModellingRule.None) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadAlreadyExists));
        }
        if (modellingRule === UaModellingRule.None) {
            return;
        }

        this._modellingRule = modellingRule;

        if (modellingRule === UaModellingRule.Mandatory) {
            this.addReference(new UaReference(UaObjects.ModellingRule_Mandatory, UaReferenceTypes.HasModellingRule, true));
        }
    }
}